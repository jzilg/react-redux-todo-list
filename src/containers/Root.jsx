import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { Provider, connect } from 'react-redux'
import Loading from '../components/Loading'
import Error from '../components/Error'
import App from './App'

const Root = ({ store, error, isLoading }) => {
    const loader = isLoading ? <Loading /> : null
    const content = error.get('appears') ? <Error message={error.get('message')} /> : <App />

    return (
        <Provider store={store}>
            <Fragment>
                {loader}
                {content}
            </Fragment>
        </Provider>
    )
}

Root.propTypes = {
    store: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
    error: ImmutablePropTypes.mapContains({
        appears: PropTypes.bool.isRequired,
        message: PropTypes.string.isRequired,
    }).isRequired,
}

const mapStateToProps = state => ({
    isLoading: state.app.get('isLoading'),
    error: state.app.get('error'),
})

export default connect(mapStateToProps)(Root)
