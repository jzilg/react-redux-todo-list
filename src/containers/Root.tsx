import React, { Fragment, ReactElement } from 'react'
import { Store } from 'redux'
import { Provider, connect } from 'react-redux'
import Notification from '../entities/notification.interface'
import ReduxState from '../redux/interfaces/state.interface'
import { unsetNotification } from '../redux/actions/ui.actions'
import Loading from '../components/loading'
import Notifications from '../components/notifications'
import App from './App'

interface Props extends StateProps {
    store: Store
}

const Root = ({ store, isLoading, notifications }: Props): ReactElement<{}> => {
    const removeNotification = (id, triggeredBy): void => {
        store.dispatch(unsetNotification(id, triggeredBy))
    }
    const loader = isLoading ? <Loading /> : null
    const notificationsElement = (
        <Notifications
            notifications={notifications}
            removeNotification={removeNotification}
        />
    )

    return (
        <Provider store={store}>
            <Fragment>
                {loader}
                {notificationsElement}
                <App />
            </Fragment>
        </Provider>
    )
}

interface StateProps {
    isLoading: boolean
    notifications: Notification[]
}

const mapStateToProps = (state: ReduxState): StateProps => ({
    isLoading: state.ui.isLoading,
    notifications: state.ui.notifications,
})

export default connect(mapStateToProps)(Root)
