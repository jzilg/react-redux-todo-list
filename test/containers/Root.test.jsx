import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { createMockStore } from 'redux-test-utils'
import { shallowToJson } from 'enzyme-to-json'
import Root from '../../src/containers/Root'

Enzyme.configure({ adapter: new Adapter() })

describe('Root', () => {
    it('should render correctly when loading', () => {
        const state = {
            ui: {
                isLoading: true,
                error: {
                    hasOccurred: false,
                    message: 'error',
                },
            },
        }
        const store = createMockStore(state)
        const output = shallow(<Root store={store} />)
        output.dive()
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should render correctly when error hasOccurred', () => {
        const state = {
            ui: {
                isLoading: false,
                error: {
                    hasOccurred: true,
                    message: 'error',
                },
            },
        }
        const store = createMockStore(state)
        const output = shallow(<Root store={store} />)
        output.dive()
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should render correctly when not loading and no error hasOccurred', () => {
        const state = {
            ui: {
                isLoading: false,
                error: {
                    hasOccurred: false,
                    message: 'error',
                },
            },
        }
        const store = createMockStore(state)
        const output = shallow(<Root store={store} />)
        output.dive()
        expect(shallowToJson(output)).toMatchSnapshot()
    })
})
