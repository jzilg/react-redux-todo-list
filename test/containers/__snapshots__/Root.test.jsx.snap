// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root should render correctly when error hasOccurred 1`] = `
<Root
  dispatch={[Function]}
  isLoading={false}
  store={
    Object {
      "dispatch": [Function],
      "getAction": [Function],
      "getActions": [Function],
      "getState": [Function],
      "isActionDispatched": [Function],
      "isActionTypeDispatched": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": undefined,
    }
  }
/>
`;

exports[`Root should render correctly when loading 1`] = `
<Root
  dispatch={[Function]}
  isLoading={true}
  store={
    Object {
      "dispatch": [Function],
      "getAction": [Function],
      "getActions": [Function],
      "getState": [Function],
      "isActionDispatched": [Function],
      "isActionTypeDispatched": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": undefined,
    }
  }
/>
`;

exports[`Root should render correctly when not loading and no error hasOccurred 1`] = `
<Root
  dispatch={[Function]}
  isLoading={false}
  store={
    Object {
      "dispatch": [Function],
      "getAction": [Function],
      "getActions": [Function],
      "getState": [Function],
      "isActionDispatched": [Function],
      "isActionTypeDispatched": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": undefined,
    }
  }
/>
`;
